ui/src/groups/AddGroupSheet/JoinGroup.tsx

```diff
@@ -0,0 +1,191 @@
+import { useRef, useState } from 'react';
+import cn from 'classnames';
+import { isValidPatp } from 'urbit-ob';
+import { preSig, whomIsFlag } from '@/logic/utils';
+import { useGangs, useGroupIndex, useGroups } from '@/state/groups';
+import { Gangs } from '@/types/groups';
+import LargeTextInput from '@/components/FullsizeTextInput';
+import CaretLeftIcon from '@/components/icons/CaretLeftIcon';
+import { ShipGroupsDisplay, ShipSearchResultsDisplay } from './SearchResults';
+import { MobileGroupPreview } from '../Join/JoinGroupModal';
+import useShipSearch from './useShipSearch';
+
+function JoinSelector(props: {
+  selectGroup: (flag: string) => void;
+  back: () => void;
+  className?: string;
+}) {
+  const [selected, setSelected] = useState('');
+  const [input, setInput] = useState('');
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  const inputIsShortcode = input.includes('/');
+  const [rawShip, name] = input?.split('/') ?? [];
+  const potentialShip =
+    rawShip && isValidPatp(preSig(rawShip)) ? preSig(rawShip) : '';
+  const potentialFlag =
+    potentialShip && inputIsShortcode ? preSig(`${potentialShip}/${name}`) : '';
+
+  const existingGangs = useGangs();
+  const groups = useGroups();
+  const { groupIndex, fetchStatus } = useGroupIndex(
+    selected || potentialShip || ''
+  );
+  const shipSearchResults = useShipSearch(input);
+
+  const indexedGangs = groupIndex
+    ? Object.entries(groupIndex)
+        .filter(([flag, preview]) => {
+          // Hide secret gangs
+          if ('afar' in preview.cordon) {
+            return false;
+          }
+
+          // if searching for shortcode, filter out gangs that don't match
+          if (inputIsShortcode) {
+            return flag.startsWith(potentialFlag);
+          }
+
+          return true;
+        })
+        .reduce(
+          (memo, [flag, preview]) => ({
+            ...memo,
+            [flag]: {
+              preview,
+              invite: null,
+              claim: flag in existingGangs ? existingGangs[flag].claim : null,
+            },
+          }),
+          {} as Gangs
+        )
+    : null;
+
+  // only consider search results where theres something meaningful to go off of
+  const validQuery = input.length && input !== '~';
+
+  const showShipSearch = !selected && !inputIsShortcode && validQuery;
+  const showGroups = selected || inputIsShortcode;
+
+  const onChange = (newInput: string) => {
+    if (whomIsFlag(newInput) && (existingGangs[newInput] || groups[newInput])) {
+      // if input matches a shortcode we already have, go to it immediately
+      setInput('');
+      props.selectGroup(newInput);
+    } else {
+      setInput(newInput);
+    }
+  };
+
+  const select = (patp: string, nickname?: string) => {
+    setInput(`Groups hosted by ${nickname || patp}`);
```

Comment:
I find this to be a little confusing because if you actually typed this into the input, you'd get no results. But ignore me if the design team has reviewed this already 🙂

Reply:
Heard. I didn't spend too much time on this since the mocks were ambiguous for this step. 

I'd favor this screen feeling "searchy" since that's where we're headed with it in the future, but can fall back to the multi select oriented input that already exists if so desired.

Ideally, I think we want affordances for 1) indicating the the result groups are all hosted by the selected ship and 2) whether or not that ship is reachable (maybe only display something if it seems offline)